### PR TITLE
Extract NPQValidationData callback to service

### PR DESCRIPTION
**Pull request by asmega**
### Context

- This is the second attempt after https://github.com/DFE-Digital/early-careers-framework/pull/863 due to other shifting changes
- There is a callback `NPQValidationData#update_participant_profile`
- It creates and persists multiple database records
- It needs to be removed so we can have finer grain control of when it happens and the behaviour of it

### Changes proposed in this pull request

- Remove the callback `NPQValidationData#update_participant_profile`
- Extract the contents to a new service `NPQ::CreateOrUpdateProfile`
- Call this new service in the api controller and in tests where the data is depended upon

### Guidance to review

- There should be no behavioural changes for existing use cases
- One would be required to call this service class if `NPQValidationData` is mutated

### Testing

- Not sure

### Review Checks
- [x] All pages have automated accessibility checks via cypress
- [x] All pages have visual tests via cypress + percy
- [x] All `School` queries are correctly scoped - `eligible` when they need to be

### How can I view this in a review app?

- Use api calls to submit NPQ registrations
- `NPQValidationData`, `TeacherProfile`, `ParticipantProfile::NPQ` should all get persisted correctly

### CODE REVIEW

File: spec/requests/api/v1/npq_applications_spec.rb [hunk 1]
```diff
@@ -229,19 +234,24 @@
     end
 
     context "when participant has applied for multiple NPQs" do
-      let!(:other_npq_profile) { create(:npq_validation_data, npq_lead_provider: npq_lead_provider, user: user) }
-      let!(:other_accepted_npq_profile) { create(:npq_validation_data, npq_lead_provider: npq_lead_provider, user: user, lead_provider_approval_status: "accepted") }
+      let!(:other_npq_validation_data) { create(:npq_validation_data, npq_lead_provider: npq_lead_provider, user: user) }
+      let!(:other_accepted_npq_validation_data) { create(:npq_validation_data, npq_lead_provider: npq_lead_provider, user: user, lead_provider_approval_status: "accepted") }
+
+      before do
+        NPQ::CreateOrUpdateProfile.new(npq_validation_data: other_npq_validation_data).call
```
Comment (kicferk1) [inline spec/requests/api/v1/npq_applications_spec.rb:241]: I like using services to set up test data, factories are cool and all, but this is nicer imo.

File: spec/requests/api/v1/npq_applications_spec.rb [hunk 2]
```diff
@@ -229,19 +234,24 @@
     end
 
     context "when participant has applied for multiple NPQs" do
-      let!(:other_npq_profile) { create(:npq_validation_data, npq_lead_provider: npq_lead_provider, user: user) }
-      let!(:other_accepted_npq_profile) { create(:npq_validation_data, npq_lead_provider: npq_lead_provider, user: user, lead_provider_approval_status: "accepted") }
+      let!(:other_npq_validation_data) { create(:npq_validation_data, npq_lead_provider: npq_lead_provider, user: user) }
+      let!(:other_accepted_npq_validation_data) { create(:npq_validation_data, npq_lead_provider: npq_lead_provider, user: user, lead_provider_approval_status: "accepted") }
```
Comment (kicferk1) [inline spec/requests/api/v1/npq_applications_spec.rb:238]: Good rename 👍

**Review by kicferk1 — APPROVED** [review]

File: spec/features/participant-declarations/submit_participant_declarations_spec.rb [hunk 3]
```diff
@@ -105,9 +105,12 @@ def given_an_npq_participant_has_been_entered_onto_the_dfe_service
     create(:schedule, name: "ECF September standard 2021")
     npq_lead_provider = create(:npq_lead_provider, cpd_lead_provider: @cpd_lead_provider)
     npq_course = create(:npq_course, identifier: "npq-leading-teaching")
-    npq_validation_date = create(:npq_validation_data, npq_lead_provider: npq_lead_provider, npq_course: npq_course)
-    @npq_id = npq_validation_date.user.id
-    travel_to npq_validation_date.profile.schedule.milestones.first.start_date + 1.day
+    npq_validation_data = create(:npq_validation_data, npq_lead_provider: npq_lead_provider, npq_course: npq_course)
+    @npq_id = npq_validation_data.user.id
+
+    NPQ::CreateOrUpdateProfile.new(npq_validation_data: npq_validation_data).call
```
Comment (kicferk1) [inline spec/features/participant-declarations/submit_participant_declarations_spec.rb:111]: I wonder if that will result in random percy failures - it seems to be adding faker-looking data to participant details that we show.

Comment (asmega) [reply spec/features/participant-declarations/submit_participant_declarations_spec.rb:111]: There are failures. I don't actually understand Percy. mainline has data that should be present missing

Comment (kicferk1) [reply spec/features/participant-declarations/submit_participant_declarations_spec.rb:111]: Huh, it seems the percy snapshots which changed are from percy, `spec/cypress/integration/admin/ParticipantManagment.feature`. Which runs a scenario you didn't touch at all in this PR.

I think removing `after_save :update_participant_profile` hook from npq_validation_data means that the syncing logic is not run in the scenario, and that surfaces Faker randomness. Uhh. 

`spec/cypress/app_commands/scenarios/admin/school_participants.rb` is the scenario in question. I think you want to change 

```
npq_profile = FactoryBot.create :participant_profile,
                                :npq,
                                validation_data: nil,
                                user: npq_user,
                                created_at: Date.parse("19/09/2019")

FactoryBot.create :npq_validation_data,
                  id: npq_profile.id,
                  user: npq_user,
                  date_of_birth: Date.parse("10/12/1982"),
                  nino: "NI123456",
                  created_at: Date.parse("19/09/2019"),
                  teacher_reference_number: "9780824"
```
to be something like 
```
validation_data = FactoryBot.create :npq_validation_data,
                  id: npq_profile.id,
                  user: npq_user,
                  date_of_birth: Date.parse("10/12/1982"),
                  nino: "NI123456",
                  created_at: Date.parse("19/09/2019"),
                  teacher_reference_number: "9780824",
                  urn: "123456"
NPQ::CreateOrUpdateProfile.new(npq_validation_data: validation_data).call
```

Comment (asmega) [reply spec/features/participant-declarations/submit_participant_declarations_spec.rb:111]: Ok, I've stuck another commit on the end roughly based on your suggestion

File: spec/cypress/app_commands/scenarios/admin/school_participants.rb [hunk 4]
```diff
@@ -47,18 +47,13 @@
...
-                  id: npq_profile.id,
-                  user: npq_user,
-                  date_of_birth: Date.parse("10/12/1982"),
-                  nino: "NI123456",
-                  created_at: Date.parse("19/09/2019"),
-                  teacher_reference_number: "9780824"
-
-npq_profile.update!(updated_at: Date.parse("19/09/2019"))
+npq_validation_data = FactoryBot.create :npq_validation_data,
+                                        id: npq_profile.id,
```
Comment (kicferk1) [inline spec/cypress/app_commands/scenarios/admin/school_participants.rb:51]: My bad, after removing npq profile above we should probably remove this line and let the db auto-generate it. 

```
Body: Puma caught this error: undefined local variable or method `npq_profile' for CypressOnRails::CommandExecutor:Class (NameError)
```
is failing :(

Comment (asmega) [reply spec/cypress/app_commands/scenarios/admin/school_participants.rb]: ok should be sorted on next run. nah, it's my fault for not checking

**Review by kicferk1 — APPROVED** [review]

**Pull request merged**